src/Microsoft.Framework.Logging/LoggerStructureFormat.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+
+namespace Microsoft.Framework.Logging
+{
+    /// <summary>
+    /// LoggerStructure to enable formatting options supported by <see cref="string.Format"/>. 
+    /// This also enables using {NamedformatItem} in the format string.
+    /// </summary>
+    public class LoggerStructureFormat : ILoggerStructure
+    {
+        private static ConcurrentDictionary<string, Formatter> _formatters = new ConcurrentDictionary<string, Formatter>();
+        private readonly Formatter _formatter;
+        private readonly object[] _values;
+
+        public LoggerStructureFormat(string format, object[] values)
+        {
+            _formatter = _formatters.GetOrAdd(format, f => new Formatter(f));
+            _values = values;
+        }
+
+        public string Message { get { return null; } }
+
+        public string Format()
+        {
+            return _formatter.Format(_values);
+        }
+
+        public IEnumerable<KeyValuePair<string, object>> GetValues()
+        {
+            return _formatter.GetValues(_values);
+        }
+
+        private class Formatter
+        {
+            private readonly string _format;
+            private readonly List<string> _valueNames = new List<string>();
+
+            public Formatter(string format)
+            {
+                var sb = new StringBuilder();
+                var scanIndex = 0;
+                var endIndex = format.Length;
+
+                while (scanIndex < endIndex)
+                {
+                    var openBraceIndex = FindIndexOf(format, '{', scanIndex, endIndex);
```

Comment:
I think we need to consider how to escape curly braces. Currently I think it's broken if you try to put an arbitrary curly brace in the format. The typical way to escape curly braces is to double them, i.e. `{{` and `}}`. That's pretty much what `string.Format` does and also ASP.NET Routing.